httpserver/server.go

```diff
@@ -192,3 +189,34 @@ type ServerAddrNotFoundError struct{}
 func (err ServerAddrNotFoundError) Error() string {
 	return "The http server address configuration is not set. Skipping server creation"
 }
+
+func serverGracefulShutdown(isTLSEnabled bool, svr *http.Server, lsnr net.Listener, certFile string, keyFile string) error {
+	connectionsClosed := make(chan struct{})
+	go func() {
+		sigint := make(chan os.Signal, 1)
+		signal.Notify(sigint, os.Interrupt, syscall.SIGTERM)
+		<-sigint
+		logrus.Info("shutting down ratify server...")
+		ctx, cancel := context.WithTimeout(context.Background(), 6*time.Second) // wait 6 seconds for existing connections to finish
+		defer cancel()
+		if err := svr.Shutdown(ctx); err != nil {
+			logrus.Errorf("failed to shutdown ratify server: %v", err)
+		}
+		close(connectionsClosed)
+	}()
+	if isTLSEnabled {
+		if err := svr.ServeTLS(lsnr, certFile, keyFile); err != nil && err != http.ErrServerClosed { // http.ErrServerClosed is immediately returned when server.Shutdown() is called
+			logrus.Errorf("failed to start server: %v", err)
+			return err
+		}
+	} else {
+		if err := svr.Serve(lsnr); err != nil && err != http.ErrServerClosed { // http.ErrServerClosed is immediately returned when server.Shutdown() is called
+			logrus.Errorf("failed to start server: %v", err)
+			return err
+		}
+	}
+	<-connectionsClosed // wait for server to shutdown
+	logrus.Info("ratify server shutdown complete")
+	time.Sleep(5 * time.Second) // wait for controllers/manager to shutdown
```

Comment:
nit: we can have a constant for 5, I feel we often use 2, 5, 7 numbers.

Reply:
do you think it's valuable to have a constant in this particular case given that it's a completely arbitrary wait amount? This 5 seconds is not related to our timeout amount. It's just a time period I noticed was sufficient to complete the shutdown operations before process exit.

Reply:
I see, if this is not related to another 5 seconds, I'm OK with it.